libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,893 @@
+        diagnostic::error("`x_min` must be less than `x_max`")
+          .primary(x_min->value.source)
+          .primary(x_max->value.source)
+          .emit(dh);
+        return failure::promise();
+      }
+    }
+    if (y_min and y_max) {
+      if (y_min->inner.get_data().index() != y_max->inner.get_data().index()) {
+        diagnostic::error("`y_min` and `y_max` must have the same type")
+          .primary(*y_min)
+          .primary(*y_max)
+          .emit(dh);
+        return failure::promise();
+      }
+      if (y_min->inner >= y_max->inner) {
+        diagnostic::error("`y_min` must be less than `y_max`")
+          .primary(*y_min)
+          .primary(*y_max)
+          .emit(dh);
+        return failure::promise();
+      }
+    }
+    if (limit.inner > 100'000) {
+      diagnostic::error("`limit` must be less than 100k")
+        .primary(limit)
+        .emit(dh);
+      return failure::promise();
+    }
+    if (limit.inner == 0) {
+      diagnostic::error("`limit` must be positive").primary(limit).emit(dh);
+      return failure::promise();
+    }
+    return {};
+  }
+
+  auto find_plugins(session ctx) const -> plugins_map {
+    auto plugins = plugins_map{};
+    auto ident = ast::identifier{"once", location::unknown};
+    const auto entity = ast::entity{std::vector{std::move(ident)}};
+    for (const auto& [_, call] : y) {
+      if (auto ptr
+          = dynamic_cast<const aggregation_plugin*>(&ctx.reg().get(call))) {
+        plugins.emplace_back(ptr, call);
+        continue;
+      }
+      auto wrapped_call = ast::function_call{entity, {call}, call.rpar, false};
+      TENZIR_ASSERT(resolve_entities(wrapped_call, ctx));
+      auto ptr
+        = dynamic_cast<const aggregation_plugin*>(&ctx.reg().get(wrapped_call));
+      TENZIR_ASSERT(ptr);
+      plugins.emplace_back(ptr, std::move(wrapped_call));
+    }
+    return plugins;
+  }
+
+  auto make_bucket(const plugins_map& plugins, session ctx) const -> bucket {
+    auto b = bucket{};
+    for (const auto& [plugin, arg] : plugins) {
+      auto inv = aggregation_plugin::invocation{arg};
+      auto instance = plugin->make_aggregation(std::move(inv), ctx);
+      TENZIR_ASSERT(instance);
+      b.push_back(std::move(instance).unwrap());
+    }
+    return b;
+  }
+};
+
+auto to_double(data&& d) -> data {
+  return match(
+    d,
+    [](std::integral auto& d) -> data {
+      return static_cast<double>(d);
+    },
+    [](auto& v) -> data {
+      return std::move(v);
+    });
+}
+
+class chart_operator2 final : public crtp_operator<chart_operator2> {
+public:
+  chart_operator2() = default;
+
+  explicit chart_operator2(chart_args args) : args_{std::move(args)} {
+  }
+
+  auto name() const -> std::string override {
+    return "tql2.chart";
+  }
+
+  auto optimize(const expression& filter, event_order) const
+    -> optimize_result override {
+    if (not args_.x_min and not args_.x_max) {
+      return {filter, event_order::unordered, copy()};
+    }
```

Comment:
This will pass the filter through, which would still be invalid. You can return `std::nullopt` here to prevent the incoming predicate from being pushed.

Reply:
Changed.